[dtensor] add op support for memory efficient attention

tianyu-l opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122997
* __->__ #122996
* #122995

This is a followup to flash attention. On cuda, flash attention is supported only for fp16/bf16, whereas memory efficient attention is supported for fp32 (but not fp64). With this PR, one can run SDPA and in general Transformer completely in dtensor.


cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @wconstab @yf225 @chauhang @d4l3k @rohan-varma

tianyu-l committed:
[dtensor] add op support for memory efficient attention

[ghstack-poisoned]

XilunWu reviewed (approved):
this looks good to me.

tianyu-l committed:
Update on "[dtensor] add op support for memory efficient attention"


This is a followup to flash attention. On cuda, flash attention is supported only for fp16/bf16, whereas memory efficient attention is supported for fp32 (but not fp64). With this PR, one can run SDPA and in general Transformer completely in dtensor.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225 chauhang

[ghstack-poisoned]

tianyu-l committed:
Update on "[dtensor] add op support for memory efficient attention"


This is a followup to flash attention. On cuda, flash attention is supported only for fp16/bf16, whereas memory efficient attention is supported for fp32 (but not fp64). With this PR, one can run SDPA and in general Transformer completely in dtensor.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225 chauhang

[ghstack-poisoned]

tianyu-l committed:
Update on "[dtensor] add op support for memory efficient attention"


This is a followup to flash attention. On cuda, flash attention is supported only for fp16/bf16, whereas memory efficient attention is supported for fp32 (but not fp64). With this PR, one can run SDPA and in general Transformer completely in dtensor.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225 chauhang

[ghstack-poisoned]

tianyu-l committed:
Update on "[dtensor] add op support for memory efficient attention"


This is a followup to flash attention. On cuda, flash attention is supported only for fp16/bf16, whereas memory efficient attention is supported for fp32 (but not fp64). With this PR, one can run SDPA and in general Transformer completely in dtensor.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225 chauhang

[ghstack-poisoned]

tianyu-l committed:
Update on "[dtensor] add op support for memory efficient attention"


This is a followup to flash attention. On cuda, flash attention is supported only for fp16/bf16, whereas memory efficient attention is supported for fp32 (but not fp64). With this PR, one can run SDPA and in general Transformer completely in dtensor.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225 chauhang

[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/ops/matrix_ops.py:
```diff
@@ -374,3 +374,183 @@ def constant_pad_nd_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> OpStrateg
+
+        # second we can accept the sharding pattern of tensor parallelism, which
+        # shard on the heads dimension
+        qkv_sharding = Shard(1)
```
is there a way to unify the common part of flash attention op strategy  and efficeint attention startegy to a function and call them differently to reuse the code?

tianyu-l replied:
I thought about if this is possible. It seems for flash attention and efficient attention, the sets of parameters and their orderings are very different -- apparently the interfaces are not aligned when they design the two attention backends. From my perspective, the part of code they share is very minimal.

So IMO it's better to separate them, especially considering there are efforts like ring attention (#122460) which only works on top of one of them.

wanchaol replied:
Hmm I was referring to the ring attention case, where it seems like we can use a templated attention function to encapsulate flash + memory attention https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/experimental/attention.py#L187

But anyways I think this could be a follow up improvement later

tianyu-l committed:
Update on "[dtensor] add op support for memory efficient attention"


This is a followup to flash attention. On cuda, flash attention is supported only for fp16/bf16, whereas memory efficient attention is supported for fp32 (but not fp64). With this PR, one can run SDPA and in general Transformer completely in dtensor.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225 chauhang

[ghstack-poisoned]

wanchaol reviewed (approved):
lgtm

tianyu-l commented:
@pytorchbot merge
